libs/base/System/Concurrency.idr

```diff
@@ -0,0 +1,191 @@
+module System.Concurrency
+
+-- At the moment this is pretty fundamentally tied to the Scheme RTS.
+-- Given that different back ends will have entirely different threading
+-- models, it might be unavoidable, but we might want to think about possible
+-- primitives that back ends should support.
+
+
+-- Thread mailboxes
+
+%foreign "scheme:blodwen-set-thread-data"
+prim__setThreadData : {a : Type} -> a -> PrimIO ()
+%foreign "scheme:blodwen-get-thread-data"
+prim__getThreadData : (a : Type) -> PrimIO a
+
+export
+setThreadData : HasIO io => {a : Type} -> a -> io ()
+setThreadData val = primIO (prim__setThreadData val)
+
+export
+getThreadData : HasIO io => (a : Type) -> io a
+getThreadData a = primIO (prim__getThreadData a)
+
+
+-- Mutexes
+
+export
+data Mutex : Type where [external]
+
+%foreign "scheme:blodwen-make-mutex"
+prim__makeMutex : PrimIO Mutex
+%foreign "scheme:blodwen-mutex-acquire"
+prim__mutexAcquire : Mutex -> PrimIO ()
+%foreign "scheme:blodwen-mutex-release"
+prim__mutexRelease : Mutex -> PrimIO ()
+
+||| Creates and returns a new mutex.
+export
+makeMutex : HasIO io => io Mutex
+makeMutex = primIO prim__makeMutex
+
+||| Acquires the mutex identified by `mutex`. The thread blocks until the mutex
+||| has been acquired.
+|||
+||| Mutexes are recursive in Posix threads terminology, which means that the
+||| calling thread can use mutex-acquire to (re)acquire a mutex it already has.
+||| In this case, an equal number of mutex-release calls is necessary to release
+||| the mutex.
+export
+mutexAcquire : HasIO io => Mutex -> io ()
+mutexAcquire m = primIO (prim__mutexAcquire m)
+
+||| Releases the mutex identified by `mutex`. Unpredictable behavior results if
```

Comment:
Seems like that "unpredictable behavior" was the reason why these functions lived in `Raw` module.

Reply:
That's fair. I'm of the opinion that mutexes and condition variables shouldn't be offered by Idris2 anyway, since implementing condition variables across backends is a nightmare, and so is testing them. I'd gladly remove them altogether.

Reply:
Imo, if we're offering thread support (fork, join, etc.) then we should also be offering mutexes and CVs. I know they're a pain to implement and test, but they're useful for that level of thread control, no? :)

Reply:
They're rather difficult to implement in the Racket backend, since Racket AFAIK does not offer them natively. Would you be able to contribute some tests for CVs, @CodingCellist?